vpr/src/place/placer.cpp

```diff
@@ -0,0 +1,403 @@
+
+#include "placer.h"
+
+#include <utility>
+
+#include "vtr_time.h"
+#include "draw.h"
+#include "read_place.h"
+#include "analytic_placer.h"
+#include "initial_placement.h"
+#include "concrete_timing_info.h"
+#include "verify_placement.h"
+#include "place_timing_update.h"
+#include "annealer.h"
+#include "RL_agent_util.h"
+#include "place_checkpoint.h"
+#include "tatum/echo_writer.hpp"
+
+Placer::Placer(const Netlist<>& net_list,
+               const t_placer_opts& placer_opts,
+               const t_analysis_opts& analysis_opts,
+               const t_noc_opts& noc_opts,
+               const std::vector<t_direct_inf>& directs,
+               std::shared_ptr<PlaceDelayModel> place_delay_model,
+               bool cube_bb,
+               bool is_flat,
+               bool quiet)
+    : placer_opts_(placer_opts)
+    , analysis_opts_(analysis_opts)
+    , noc_opts_(noc_opts)
+    , costs_(placer_opts.place_algorithm, noc_opts.noc)
+    , placer_state_(placer_opts.place_algorithm.is_timing_driven(), cube_bb)
+    , rng_(placer_opts.seed)
+    , net_cost_handler_(placer_opts, placer_state_, cube_bb)
+    , place_delay_model_(std::move(place_delay_model))
+    , log_printer_(*this, quiet)
+    , is_flat_(is_flat) {
+    const auto& cluster_ctx = g_vpr_ctx.clustering();
+    const auto& device_ctx = g_vpr_ctx.device();
+    const auto& atom_ctx = g_vpr_ctx.atom();
+
+    const auto& timing_ctx = g_vpr_ctx.timing();
+    pre_place_timing_stats_ = timing_ctx.stats;
+
+    init_placement_context(placer_state_.mutable_blk_loc_registry(), directs);
+
+    // create a NoC cost handler if NoC optimization is enabled
+    if (noc_opts.noc) {
+        noc_cost_handler_.emplace(placer_state_.block_locs());
+    }
+
+    // Start measuring placement time
+    timer_ = std::make_unique<vtr::ScopedStartFinishTimer>("Placement");
+    timer_->quiet(quiet);
+
+    /* To make sure the importance of NoC-related cost terms compared to
+     * BB and timing cost is determine only through NoC placement weighting factor,
+     * we normalize NoC-related cost weighting factors so that they add up to 1.
+     * With this normalization, NoC-related cost weighting factors only determine
+     * the relative importance of NoC cost terms with respect to each other, while
+     * the importance of total NoC cost to conventional placement cost is determined
+     * by NoC placement weighting factor.
+     */
+    if (noc_opts.noc) {
+        normalize_noc_cost_weighting_factor(const_cast<t_noc_opts&>(noc_opts));
+    }
+
+    BlkLocRegistry& blk_loc_registry = placer_state_.mutable_blk_loc_registry();
+    initial_placement(placer_opts, placer_opts.constraints_file.c_str(),
```

Comment:
Could pass a block location registry to a separate update_locations (or some such) function instead if you wished. But I'm OK with this if it serves the needs of everybody.